pretyping/evarsolve.ml

```diff
@@ -73,25 +72,33 @@ let refresh_universes ?(status=univ_rigid) ?(onlyalg=false) ?(refreshset=false)
        refresh_sort status ~direction s
       | _ -> t
       end
-    | Prod (na,u,v) -> 
-      mkProd (na, u, refresh ~onlyalg status ~direction v)
+    | Prod (na,u,v) ->
+       let v' = refresh ~onlyalg status ~direction v in
+       if v' == v then t else mkProd (na, u, v')
     | _ -> t
   (** Refresh the types of evars under template polymorphic references *)
   and refresh_term_evars onevars top t =
     match EConstr.kind !evdref t with
     | App (f, args) when is_template_polymorphic env !evdref f ->
       let pos = get_polymorphic_positions !evdref f in
-	refresh_polymorphic_positions args pos
+        refresh_polymorphic_positions args pos; t
     | App (f, args) when top && isEvar !evdref f -> 
-      refresh_term_evars true false f; 
-      Array.iter (refresh_term_evars onevars false) args
+       let f' = refresh_term_evars true false f in
+       let args' = Array.map (refresh_term_evars onevars false) args in
+       if f' == f && args' == args then t
+       else mkApp (f', args')
     | Evar (ev, a) when onevars ->
```

Comment:
What happens if the `Evar` has no instantiation, but candidates containing algebraic universes? Can that happen?

Reply:
It can happen but it is fine. We are only interested to ensure that the declared type of the evar is refreshed.

Reply:
@mattam82 answered on gitter that if resolution by a candidate uses `evar_define`, it should be fine.

Reply:
Oops! Interesting race condition :)